### ISSUE TITLE
Large Batch Simulation Breaks with Mixed Precision

### ISSUE THREAD

**Issue by czmrand** [horovod/horovod#2957]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 2.3
3. Horovod version: 0.22.0
7. Python version: 3.7

**Checklist:**
1. Did you search issues to find if somebody asked this question before? yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Bug report:**
Please describe erroneous behavior you're observing and steps to reproduce it.

When running large batch simulation with mixed precision the accumulated gradients cannot recover from a nan gradient. As a result, training will get stuck.
The reason is the use of the following code to reset the aggregated grads:

    def _clear_vars(self):
        self.counter.assign(0)
        for idx in self.locally_aggregated_grads.keys():
            self.locally_aggregated_grads[idx].assign(
                -1 * self.locally_aggregated_grads[idx])

I have fixed this by changing the code to:

    def _clear_vars(self):
        self.counter.assign(0)
        for idx in self.locally_aggregated_grads.keys():
            self.locally_aggregated_grads[idx].assign(
                tf.zeros_like(self.locally_aggregated_grads[idx]))

**Comment by czmrand** [comment]
I have found a few additional issues when using TensorFlow mixed precision with gradient aggregator.
1.
It seems that starting from TensorFlow 2.4, the gradient aggregation is performed inside the apply_gradients function. This is incompatible with mixed precision due to the fact that the LossScaleOptimizer does not always call apply_gradients.
To fix this I have modified the code to perform the aggregation before the apply_gradients call (similar to pre tf 2.4) and removed the use of the _POST_TF_2_4_0 flag:
In the Horovod optimizer I override the _compute_gradients optimizer function and have it call the following function in LocalGradientAggregationHelperEager (which is nearly identical to the compute_gradients function)

    def accumulate_gradients(self, grads_and_vars):
        # On steps where allreduce happens, resulting_grads returns the allreduced
        # gradients, on other steps it returns the locally aggregated
        # gradients.
        grads, vars = list(zip(*grads_and_vars))
        resulting_grads = []

        for idx, grad in enumerate(grads):
            # Handle IndexedSlices.
            if self.sparse_as_dense and isinstance(grad, tf.IndexedSlices):
                grad = tf.convert_to_tensor(grad)
            elif isinstance(grad, tf.IndexedSlices):
                raise ValueError(
                    "IndexedSlices are not supported when "
                    "`backward_passes_per_step` > 1 and "
                    "`sparse_as_dense` is False."
                )

            # Create variables to store to aggregate gradients if they don't
            # already exist. Skip variables that are None.
            if idx not in self.locally_aggregated_grads.keys():
                if grad is not None:
                    self.locally_aggregated_grads[idx] = tf.Variable(
                        initial_value=tf.zeros_like(grad),
                        trainable=False,
                        dtype=grad.dtype,
                    )

            if grad is None:
                resulting_grads.append(None)
            else:
                self.locally_aggregated_grads[idx].assign_add(grad)
                resulting_grads.append(
                    self.locally_aggregated_grads[idx].read_value())
        assert len(self.locally_aggregated_grads) == len(grads)

        # Increment counter.
        self.counter.assign_add(1)

        def _all_reduce_and_clear_aggregated_variables(aggregated_gradients):
            # Performs allreduce. If `average_aggregated_gradients` is
            # set to True divides result by `backward_passes_per_step`.
            reduced_gradients = self._allreduce_helper(aggregated_gradients)
            assert len(reduced_gradients) == len(grads)

            self._clear_vars()
            return reduced_gradients

        def _do_nothing(aggregated_gradients):
            return aggregated_gradients

        resulting_grads = tf.cond(
            pred=tf.equal(self.counter, self.backward_passes_per_step),
            true_fn=lambda: _all_reduce_and_clear_aggregated_variables(resulting_grads),
            false_fn=lambda: _do_nothing(resulting_grads),
        )
        return list(zip(resulting_grads, vars))

2. If you use a dynamic loss scaler, you need to make sure that scale updates only occur on steps in which the gradient aggregation is reset. This is because the gradients are aggregated in their scaled state and if they were calculated with different scales the aggregation will be wrong. Here is one hacky way to fix:
In the loss scale optimizer when checking whether to update the scaler, check the internal state of the optimizer:
    if hasattr(self._optimizer,'_agg_helper') and self._optimizer._agg_helper is not None:
       import tensorflow as tf
       should_apply_grads = tf.equal(self._optimizer._agg_helper.counter, 0)
       loss_scale_update_op, should_apply_grads = tf.cond(
            pred=should_apply_grads,
            true_fn=lambda: self._loss_scale.update(grads),
            false_fn=lambda: (tf.no_op(),should_apply_grads) ,
        )
    else:
        loss_scale_update_op, should_apply_grads = self._loss_scale.update(grads)

**Comment by romerojosh** [comment]
Hi @czmrand, thanks for the investigation. Do you think you'd be able to put together a PR with these fixes for TF 2.4?
